File: mirror/tier2-dart/showcase/meta_reflection.dart
// MIRROR V3: Tier 2 Dart Showcase
// Category: Metaprogramming
// Module: meta_reflection.dart

import 'dart:mirrors';

Map<String, dynamic> reflectObject(dynamic obj) {
  return <String, dynamic>{
    'type': obj.runtimeType.toString(),
    'instanceMethods': 'available',
  };
}

Map<String, dynamic> reflectFunction(Function func) {
  return <String, dynamic>{
    'callable': true,
    'reflection': 'available',
  };
}

Map<String, dynamic> getAttributes(dynamic obj) {
  final attrs = <String, String>{};
  if (obj is Map) {
    for (final entry in (obj as Map).entries) {
      attrs[entry.key.toString()] = entry.value.runtimeType.toString();
    }
  }
  return attrs.cast<String, dynamic>();
}

class ReflectedClass {
  final dynamic x;
  final dynamic y;
  
  ReflectedClass(this.x, this.y);
  
  String method() => '$x:$y';
}

void main() {
  final result = <String, dynamic>{
    'reflectInt': reflectObject(42),
    'reflectStr': reflectObject('hello'),
    'reflectList': reflectObject([1, 2, 3]),
    'classAttrs': 2,
  };
  print(result);
}

File: mirror/tier2-ruby/showcase/meta_reflection.rb
# MIRROR V3: Tier 2 Ruby Showcase
# Category: Metaprogramming
# Module: meta_reflection.rb
# Purpose: Demonstrate reflection and introspection

def reflect_object(obj)
  {
    type: obj.class.name,
    class_name: obj.class.to_s,
    methods_count: obj.methods.size,
  }
end

def reflect_function(func)
  {
    name: func.class.name,
    arity: func.arity,
    callable: func.respond_to?(:call),
  }
end

def get_attributes(obj)
  attrs = {}
  obj.instance_variables.each do |var|
    attrs[var.to_s] = obj.instance_variable_get(var).class.name
  end
  attrs
end

class ReflectedClass
  def initialize(x, y)
    @x = x
    @y = y
  end
  
  def method
    "#{@x}:#{@y}"
  end
end

result = {
  reflect_int: reflect_object(42),
  reflect_str: reflect_object("hello"),
  reflect_list: reflect_object([1, 2, 3]),
  class_attrs: get_attributes(ReflectedClass.new(1, "a")).size,
}

puts result.inspect
